modules/face/test/test_facemark_lbf.cpp

```diff
@@ -0,0 +1,168 @@
+ export OPENCV_TEST_DATA_PATH=/home/opencv/opencv_extra/testdata
+ <build_folder>/bin/opencv_test_face
+*/
+
+#include "test_precomp.hpp"
+#include "opencv2/imgcodecs.hpp"
+#include "opencv2/face.hpp"
+#include <vector>
+#include <string>
+using namespace std;
+using namespace cv;
+using namespace cv::face;
+
+CascadeClassifier cascade_detector;
+bool myCustomDetector( InputArray image, OutputArray ROIs, void * config = 0 ){
+    Mat gray;
+    std::vector<Rect> & faces = *(std::vector<Rect>*) ROIs.getObj();
+    faces.clear();
+
+    if(config!=0){
+        //do nothing
+    }
+
+    if(image.channels()>1){
+        cvtColor(image.getMat(),gray,CV_BGR2GRAY);
+    }else{
+        gray = image.getMat().clone();
+    }
+    equalizeHist( gray, gray );
+
+    cascade_detector.detectMultiScale( gray, faces, 1.4, 2, CV_HAAR_SCALE_IMAGE, Size(30, 30) );
+    return true;
+}
+
+TEST(CV_Face_FacemarkLBF, can_create_default) {
+    FacemarkLBF::Params params;
+    params.n_landmarks = 68;
+
+    Ptr<Facemark> facemark;
+    EXPECT_NO_THROW(facemark = FacemarkLBF::create(params));
+    EXPECT_FALSE(facemark.empty());
+}
+
+TEST(CV_Face_FacemarkLBF, can_set_custom_detector) {
+    string cascade_filename =
+        cvtest::findDataFile("cascadeandhog/cascades/lbpcascade_frontalface.xml", true);
+
+    FacemarkLBF::Params params;
+    params.cascade_face = cascade_filename;
+
+    EXPECT_TRUE(cascade_detector.load(cascade_filename));
+
+    Ptr<Facemark> facemark = FacemarkLBF::create(params);
+    EXPECT_TRUE(facemark->setFaceDetector(myCustomDetector));
+}
+
+TEST(CV_Face_FacemarkLBF, can_perform_training) {
+
+    string i1 = cvtest::findDataFile("face/therock.jpg", true);
+    string p1 = cvtest::findDataFile("face/therock.pts", true);
+    string i2 = cvtest::findDataFile("face/mark.jpg", true);
+    string p2 = cvtest::findDataFile("face/mark.pts", true);
+
+    std::vector<string> images_train;
+    images_train.push_back(i1);
+    images_train.push_back(i2);
+
+    std::vector<String> points_train;
+    points_train.push_back(p1);
+    points_train.push_back(p2);
+
+    string cascade_filename =
+        cvtest::findDataFile("cascadeandhog/cascades/lbpcascade_frontalface.xml", true);
+    FacemarkLBF::Params params;
+    params.cascade_face = cascade_filename;
+    params.verbose = false;
+    Ptr<Facemark> facemark = FacemarkLBF::create(params);
+
+    Mat image;
+    std::vector<Point2f> landmarks;
+    for(size_t i=0;i<images_train.size();i++){
+        image = imread(images_train[i].c_str());
+        EXPECT_TRUE(loadFacePoints(points_train[i].c_str(),landmarks));
+        EXPECT_TRUE(landmarks.size()>0);
+        EXPECT_TRUE(facemark->addTrainingSample(image, landmarks));
+    }
+
+    EXPECT_NO_THROW(facemark->training());
+}
+
+TEST(CV_Face_FacemarkLBF, can_load_model) {
+    string model_filename = "LBF.model";
+
+    Ptr<Facemark> facemark = FacemarkLBF::create();
+    EXPECT_NO_THROW(facemark->loadModel(model_filename.c_str()));
```

Comment:
Each test should be standalone and don't rely on results of other tests.
Also tests should not write any data (models) if it is not required.

Probably we should put model somewhere or we can merge train/load tests into one.

Reply:
merging the train-load test is easier, i can put a flag for not saving the model.